File: apps/web/lib/recall-queue.ts
export type StudyRating = "right" | "wrong";

export type StudyMeta = {
  /** Wrong answers this session (for leave-round SM-2 flush). */
  wrongAttempts: number;
};

export const EMPTY_STUDY_META: StudyMeta = {
  wrongAttempts: 0,
};

export type StudyStep = {
  submit: 0 | 4 | null;
  done: boolean;
  meta: StudyMeta;
};

/** Session hopper: Right parks the card; Wrong sends it to the back of the queue. */
export function applyStudy(meta: StudyMeta, rating: StudyRating): StudyStep {
  if (rating === "right") {
    return { submit: 4, done: true, meta };
  }
  return {
    submit: null,
    done: false,
    meta: { wrongAttempts: meta.wrongAttempts + 1 },
  };
}

/** Failed round: write a miss if the learner marked Wrong but never got Right. */
export function leaveQuality(meta: StudyMeta): 0 | null {
  if (meta.wrongAttempts > 0) return 0;
  return null;
}

/** @deprecated Use StudyRating */
export type RecallRating = StudyRating;
/** @deprecated Use StudyMeta */
export type RecallMeta = StudyMeta;
/** @deprecated Use EMPTY_STUDY_META */
export const EMPTY_RECALL_META = EMPTY_STUDY_META;
/** @deprecated Use StudyStep */
export type RecallStep = StudyStep;
/** @deprecated Use applyStudy */
export const applyRecall = applyStudy;

File: apps/ios/Kiri/Services/RecallQueue.swift
import Foundation

/// Keep in sync with apps/web/lib/recall-queue.ts
enum StudyRating {
    case right, wrong
}

struct StudyMeta {
    var wrongAttempts: Int = 0
}

struct StudyStep {
    let submit: Int?
    let done: Bool
    let meta: StudyMeta
}

enum StudyQueue {
    static func applyStudy(meta: StudyMeta, rating: StudyRating) -> StudyStep {
        switch rating {
        case .right:
            return StudyStep(submit: 4, done: true, meta: meta)
        case .wrong:
            return StudyStep(
                submit: nil,
                done: false,
                meta: StudyMeta(wrongAttempts: meta.wrongAttempts + 1)
            )
        }
    }

    static func leaveQuality(meta: StudyMeta) -> Int? {
        meta.wrongAttempts > 0 ? 0 : nil
    }
}

struct StudyQueueItem: Identifiable {
    let id: String
    let card: RecallCardDTO
    var meta: StudyMeta
}

typealias RecallRating = StudyRating
typealias RecallMeta = StudyMeta
typealias RecallStep = StudyStep
typealias RecallQueue = StudyQueue
typealias RecallQueueItem = StudyQueueItem
